R/tbl-lazy.R

```diff
@@ -105,7 +105,60 @@ the$warned_on_tbl_lazy_names <- FALSE
 
 #' @export
 names.tbl_lazy <- function(x) {
-  # FIXME find a way to nudge users away from using `names()` without RStudio
-  # getting on your nerves
+  if (is_direct(rlang::caller_env())) {
+    cli::cli_inform(c(
+      `!` = "The {.fn names} method of {.cls tbl_lazy} is not intended to be called.",
+      i = "Did you mean {.fn colnames}?"
+    ))
+  }
   NextMethod("names")
 }
+
+#' @export
+`$.tbl_lazy` <- function(x, name) {
+  unexpected_name <- !name %in% c("lazy_query", "src")
+
+  if (unexpected_name) {
+    cli::cli_abort(c(
+      "The {.fn $} method of {.cls tbl_lazy} is not intended to be called.",
+      i = "Use {.fn dplyr::pull} to get the values in a column."
+    ))
+  }
+  NextMethod("$")
+}
+
+# COPIED FROM lifecycle
+# https://github.com/r-lib/lifecycle/blob/9417eca8f5091f95b4569fb6c388e4394e2b2157/R/deprecate.R#L358
+is_direct <- function(env) {
+  env_inherits_global(env) || from_testthat(env)
+}
```

Comment:
You can use `rlang::env_is_user_facing()`: https://rlang.r-lib.org/reference/env_is_user_facing.html 😄

Reply:
That makes is so much easier 😄